### PR TITLE
Add Introduction to PyTorch tutorial

### PR DESCRIPTION
**Pull request by adamjstewart**
The first of many tutorials introduced for #2418 

This tutorial is designed for remote sensing folks with little to no background in deep learning or PyTorch. It provides a foundation upon which all later tutorials will build.

### CODE REVIEW

File: docs/tutorials/pytorch.ipynb [hunk 1]
```diff
@@ -0,0 +1,507 @@
+{
+ "cells": [
+  {
+   "cell_type": "code",
```
Comment (adamjstewart) [inline docs/tutorials/pytorch.ipynb:4]: Would like to start formatting the copyright in code cells, as tools like `ruff` can only check code cells: https://github.com/astral-sh/ruff/issues/12802

File: docs/tutorials/pytorch.ipynb [hunk 2]
```diff
@@ -0,0 +1,507 @@
...
+  {
+   "cell_type": "markdown",
+   "id": "9478ed9a",
+   "metadata": {
+    "id": "NdrXRgjU7Zih"
+   },
+   "source": [
+    "# Introduction to PyTorch\n",
+    "\n",
+    "_Written by: Adam J. Stewart_\n",
```
Comment (adamjstewart) [inline docs/tutorials/pytorch.ipynb:23]: Would like to start giving people credit for the tutorials they write, especially for the case study tutorials. This is intended to be the first person who wrote the tutorial, we wouldn't append other names except in the case of major revisions.

File: docs/tutorials/pytorch.ipynb [hunk 3]
```diff
@@ -0,0 +1,507 @@
...
+    "In machine learning, we typically iterate over our datasets multiple times. Each full pass through the dataset is called an *epoch*. The following hyperparameter controls the number of epoch for which we train our model, and can be modified to train the model for longer:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "eb5dc7e8-6cb3-4457-83ad-7fa5aef8ea0c",
+   "metadata": {
+    "nbmake": {
+     "mock": {
```
Comment (adamjstewart) [inline docs/tutorials/pytorch.ipynb:414]: The tutorial runs for 100 epochs, but we mock it to only run for a single epoch during testing.

File: docs/tutorials/pytorch.ipynb [hunk 4]
```diff
@@ -0,0 +1,507 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "7cd0bd25-e19a-4b26-94a1-fe9a544e8afd",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "evaluate(test_dataloader)"
```
Comment (adamjstewart) [inline docs/tutorials/pytorch.ipynb:460]: Final accuracies are 65% on validation and 85% on testing. Pretty good for only 100 images. We are nowhere near plateauing, and could run for longer to get higher accuracies, but I think this proves the point.

File: docs/tutorials/pytorch.ipynb (hunk 3 above)
Comment (adamjstewart) [reply docs/tutorials/pytorch.ipynb:414]: This new tutorial only takes 9.69s to test.

**Comment by calebrob6** [comment]
Would not put "self driving cars" as an application of DL

File: docs/tutorials/pytorch.ipynb (hunk 2 above)
Comment (calebrob6) [reply docs/tutorials/pytorch.ipynb:23]: great idea

**Comment by adamjstewart** [comment]
Why not?

File: docs/tutorials/pytorch.ipynb [hunk 5]
```diff
@@ -0,0 +1,507 @@
...
+   "id": "7f26e4b8",
+   "metadata": {
+    "id": "5rLknZxrBEMz"
+   },
+   "source": [
+    "## Datasets\n",
+    "\n",
+    "In order to learn by example, we first need examples. In machine learning, we construct datasets of the form:\n",
+    "\n",
+    "$$\\mathcal{D} = \\left\\{\\left(x^{(i)}, y^{(i)}\\right)\\right\\}_{i=1}^N$$\n",
```
Comment (nilsleh) [inline docs/tutorials/pytorch.ipynb:118]: These math notations are barely visible when looking at the docs page

File: docs/tutorials/pytorch.ipynb [hunk 6]
```diff
@@ -0,0 +1,507 @@
...
+  {
+   "cell_type": "markdown",
+   "id": "f89e20ae-d3b6-4f05-a83f-f7034dd9862f",
+   "metadata": {},
+   "source": [
+    "In machine learning, we not only want to train a model, but also evaluate its performance on unseen data. Oftentimes, our dataset is split into three separate subsets:\n",
+    "\n",
+    "* train: for training the model *parameters*\n",
+    "* val: for validating the model *hyperparameters*\n",
+    "* test: for testing the model *performance*\n",
```
Comment (nilsleh) [inline docs/tutorials/pytorch.ipynb:164]: Validation is also to assess model performance right? Strictly speaking, testing data should only be run once at the very end to "deploy" the model, although almost everyone is indirectly tuning on test set performance as well.

File: docs/tutorials/pytorch.ipynb [hunk 7]
```diff
@@ -0,0 +1,507 @@
...
+   "id": "a12b9e9b-26cc-43f4-a517-31b805862df5",
+   "metadata": {},
+   "source": [
+    "## Loss Function\n",
+    "\n",
+    "If $y$ is our expected output (also called \"ground truth\") and $\\hat{y}$ is our predicted output, our goal is to minimize the difference between $y$ and $\\hat{y}$. This difference is referred to as *error* or *loss*, and the loss function tells us how big of a mistake we made. For regression tasks, a simple mean squared error is sufficient:\n",
+    "\n",
+    "$$\\mathcal{L}(y, \\hat{y}) = \\left(y - \\hat{y}\\right)^2$$\n",
+    "\n",
+    "For classification tasks, such as EuroSAT, we instead use a negative log-likelihood:\n",
```
Comment (nilsleh) [inline docs/tutorials/pytorch.ipynb:266]: MSE is also log-likelihood of a Gaussian with unit variance.

File: docs/tutorials/pytorch.ipynb (hunk 5 above)
Comment (adamjstewart) [reply docs/tutorials/pytorch.ipynb:118]: I don't think it looks too bad, the characters are of similar size to the normal text. During the tutorial, I'll likely zoom in so that everyone can see.

File: docs/tutorials/pytorch.ipynb (hunk 6 above)
Comment (adamjstewart) [reply docs/tutorials/pytorch.ipynb:164]: Technically, you're only supposed to use validation for hyperparameter tuning (and early stopping). You do evaluate performance on val, but only to decide which set of hyperparams is best. I think I've used it as such in the rest of the tutorial.

It may be worth noting that sometimes test isn't public, but I didn't go into that much detail here.

File: docs/tutorials/pytorch.ipynb (hunk 7 above)
Comment (adamjstewart) [reply docs/tutorials/pytorch.ipynb:266]: Interesting, didn't know that. I think it's easier to explain MSE as is, although it would be nice to tie into NLL/CE more smoothly.
